sympy/codegen/ast.py

```diff
@@ -399,3 +448,515 @@ def body(self):
+            super(ComplexType, self)._cast_nocheck(im(value))*1j
+        )
+
+    def _check(self, value):
+        from sympy.functions import re, im
+        super(ComplexType, self)._check(re(value))
+        super(ComplexType, self)._check(im(value))
+
+
+# NumPy types:
+intc = IntBaseType('intc')
+intp = IntBaseType('intp')
+int8 = SignedIntType('int8', 8)
+int16 = SignedIntType('int16', 16)
+int32 = SignedIntType('int32', 32)
+int64 = SignedIntType('int64', 64)
+uint8 = UnsignedIntType('uint8', 8)
+uint16 = UnsignedIntType('uint16', 16)
+uint32 = UnsignedIntType('uint32', 32)
+uint64 = UnsignedIntType('uint64', 64)
+float32 = FloatType(
+    'float32', 32,
+    max=Float('3.40282347e+38', precision=32+8),
+    tiny=Float('1.17549435e-38', precision=32+8),
+    eps=Float('1.1920929e-07', precision=32+8),
+    dig=6, decimal_dig=9
+)
+float64 = FloatType(
+    'float64', 64,
+    max=Float('1.79769313486231571e+308', precision=64+8),
+    tiny=Float('2.22507385850720138e-308', precision=64+8),
+    eps=Float('2.22044604925031308e-16', precision=64+8),
+    dig=15, decimal_dig=17
+)
+float80 = FloatType(
+    'float80', 80,
+    max=Float('1.18973149535723176502e+4932', precision=80+8),
+    tiny=Float('3.36210314311209350626e-4932', precision=80+8),
+    eps=Float('1.08420217248550443401e-19', precision=80+8),
+    dig=18, decimal_dig=21
+)
+complex64 = ComplexType('complex64', 64, **{k: getattr(float32, k) for k in FloatType.__slots__[2:]})
+complex128 = ComplexType('complex128', 128, **{k: getattr(float64, k) for k in FloatType.__slots__[2:]})
+
+# Generic types (precision may be chosen by code printers):
+real = Type('real')
+integer = IntBaseType('integer')
+complex_ = Type('complex')
+bool_ = Type('bool')
+
+
+class Attribute(Token):
+    """ Variable attribute """
+    __slots__ = ['name']
+
+value_const = Attribute('value_const')
+pointer_const = Attribute('pointer_const')
+
+class Variable(Basic):
+    """ Represents a variable
+
+    Parameters
+    ----------
+    symbol : Symbol
+    attrs : iterable of Attribute instances
+        Will be stored as a FiniteSet.
+    type_ : Type (optional)
+        Type of the variable.
+
+    Examples
+    --------
+    >>> from sympy import Symbol
+    >>> from sympy.codegen.ast import Variable, float32, integer
+    >>> x = Symbol('x')
+    >>> v = Variable(x, type_=float32)
+
+    One may also construct a ``Variable`` instance with the type deduced from
+    assumptions about the symbol using the ``deduced`` classmethod::
+    >>> i = Symbol('i', integer=True)
+    >>> v = Variable.deduced(i)
+    >>> v.type == integer
+    True
+
+    """
+
+    nargs = (2, 3)  # type is optional
+
+    def __new__(cls, symbol, attrs=None, type_=None):
+        args = (_sympify(symbol), FiniteSet() if attrs is None else FiniteSet(*attrs))
+        if type_ is not None:
+            if not isinstance(type_, Type):
+                raise NotImplementedError("Expected a Type as type_")
```

Comment:
This is a confusing error message. How about "TypeError("type_ argument should be an instance of Type")` (which is maybe also confusing...)